### PR TITLE
view info: preserve integrity by allowing base info for reads only and by initializing base info

### PR DESCRIPTION
**Pull request by eliransin**
This PR purpose is to handle schema integrity issues that can arise in races involving materialized views.
The possibility of such integrity issues was found in #7420 , where a view schema was used for reading without
it's _base_info member initialized resulting in a segfault.
We handle this doing 3 things:
1. First guard against using an uninitialized base info - this will be considered as an internal error as it will indicate that
there is a path in our code that creates a view schema to be used for reads or writes but is not initializing the base info.
2. We allow the base info to be initialized also from partially matching base (most likely a newer one that this used to create the view).
3. We fix the suspected path that create such a view schema to initialize it. (in migration manager)

It is worth mentioning that this PR is a workaround to a probable design flaw in our materialized views which requires the base
table's information to be retrieved in the first place instead of just being self contained.

Refs #7420

### CODE REVIEW

**Comment by eliransin** [comment]
@tgrabiec , @nyh to make the review easier, I will highlight the differences from the other PR which was dequeued (#7440).
1. @tgrabiec was right :slightly_smiling_face: , putting the code in the schema registry broke things, the reason is probably the 
fact  that on startup the view schemas can be registered first and only then the base schemas, this causes an exception on 
startup and eventually failure to boot. - so for short the code moved to the loader in the migration manager.
2. I have added a commit that will convert a potential null deference to an internal error. This was added because the code was moved out of the schema registry so I cant be 100% sure that this commit covers all path where this can happen.

The support for base that is used for read only stayed the same.

File: db/view/view.hh [hunk 1]
```diff
@@ -48,11 +48,30 @@ namespace view {
...
+    std::vector<column_id> _base_non_pk_columns_in_view_pk;
+public:
+    const std::vector<column_id>& base_non_pk_columns_in_view_pk() const;
+    const schema_ptr& base_schema() const;
+
+    // Indicates if the view hase pk columns which are not part of the base
+    // pk, it seems that !base_non_pk_columns_in_view_pk.empty() is the same,
+    // but actually there are cases where we can compute this boolean without
+    // succeeding to reliably build the former.
+    bool has_base_non_pk_columns_in_view_pk;
```
[the remark this answers is not in the document]
Comment (eliransin) [reply db/view/view.hh:64]: They can't change, we just build a new base info.
I will const them.

File: db/view/view.hh [hunk 2]
```diff
@@ -48,11 +48,30 @@ namespace view {
...
+
+    // If base_non_pk_columns_in_view_pk couldn't reliably be built, this base
+    // info can't be used for computing view updates, only for reading the materialized
+    // view.
+    bool use_only_for_reads;
```
[the remark this answers is not in the document]
Comment (eliransin) [reply db/view/view.hh:69]: same here.

File: schema_registry.cc [hunk 3]
```diff
@@ -24,6 +24,8 @@
 #include "schema_registry.hh"
 #include "log.hh"
 #include "db/schema_tables.hh"
+#include "database.hh"
+#include "view_info.hh"
```
Comment (tgrabiec) [inline schema_registry.cc:28]: This is now unneeded.

**Review by tgrabiec — APPROVED** [review]

File: schema_registry.cc (hunk 3 above)
Comment (eliransin) [reply schema_registry.cc]: done

**Pull request merged**